"no-invalid-this" gives false-negative result

* **ESLint Version:** 4.4.1
* **Node Version:** 8.2.1
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**What did you do? Please include the actual source code causing the issue.**

```sh
../node_modules/.bin/eslint --no-eslintrc no-invalid-this-test.js 
```

<!-- Paste the source code below: -->
**no-invalid-this-test.js** file contents:
```js
/*eslint no-invalid-this: "error"*/
/*eslint-env es6*/

let o = {
    name: "O name",
    f: function() {
        console.log(this.name)
    }
};

o.f = ((func, arg) => function() {
    console.log('Executing wrapper with arg ' + arg);
    func.call(this);
})(o.f, 'blah');

o.f();
```

**What did you expect to happen?**
No error.

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
/Users/vitaly/Work/n/build/no-invalid-this-test.js
  13:15  error  Unexpected 'this'  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

--- a/lib/ast-utils.js
+++ b/lib/ast-utils.js
@@ -626,6 +626,9 @@ module.exports = {
                 //     // setup...
                 //     return function foo() { ... };
                 //   })();
+                //   obj.foo = (() =>
+                //     function foo() { ... }
+                //   )();
                 case "ReturnStatement": {
                     const func = getUpperFunction(parent);
 
@@ -635,6 +638,12 @@ module.exports = {
                     node = func.parent;
                     break;
                 }
+                case "ArrowFunctionExpression":
+                    if (node !== parent.body || !isCallee(parent)) {
+                        return true;
+                    }
+                    node = parent.parent;
+                    break;
 
                 // e.g.
                 //   var obj = { foo() { ... } };

--- a/tests/lib/rules/no-invalid-this.js
+++ b/tests/lib/rules/no-invalid-this.js
@@ -308,6 +308,26 @@ const patterns = [
         valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
+    {
+        code: "obj.foo = (() => function() { console.log(this); z(x => console.log(x, this)); })();",
+        parserOptions: { ecmaVersion: 6 },
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
+        invalid: []
+    },
+    {
+        code: "obj.foo = (function() { return () => { console.log(this); z(x => console.log(x, this)); }; })();",
+        parserOptions: { ecmaVersion: 6 },
+        valid: [NORMAL],
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES],
+        errors
+    },
+    {
+        code: "obj.foo = (() => () => { console.log(this); z(x => console.log(x, this)); })();",
+        parserOptions: { ecmaVersion: 6 },
+        valid: [NORMAL],
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES],
+        errors
+    },
 
     // Class Instance Methods.
     {

